### PR TITLE
Added EurekaVipAddressRoundRobinWithAzAffinityService to the eureka s…

### PR DESCRIPTION
**Pull request by cjha**
…ub-module. This service will round-robin requests to instances in the same availability zone as the current server. If there are no instances in the current availability zone, it will round-robin instances in all availability zones. Routing requests this way has the advantage of lower response times and lower data transfer costs.

### CODE REVIEW

File: riposte-service-registration-eureka/src/main/java/com/nike/riposte/serviceregistration/eureka/helpers/EurekaVipAddressService.java [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+     * associated with the given VIP name (i.e. instances whose state is {@link InstanceInfo.InstanceStatus#UP}), or an
+     * empty list if no such instances were found. Will never return null.
+     * <p/>
+     * The eureka discovery client has an internal cache, so this method may return quickly, but if the cache doesn't
+     * have info on the VIP or needs to be refreshed then this method could take as long it takes the eureka discovery
+     * client to do the necessary network calls.
+     * <p/>
+     * TODO: See if Netflix has a newer version of the Eureka client that allows for async nonblocking retrieval via futures.
+     */
+    List<InstanceInfo> getActiveInstanceInfosForVipAddressBlocking(String vipName) {
```
[the remark this answers is not in the document]
Comment (cjha) [reply riposte-service-registration-eureka/src/main/java/com/nike/riposte/serviceregistration/eureka/helpers/EurekaVipAddressService.java]: Thanks for catching that. I had initially set this as an interface but switched it to an abstract class and forgot to put back in the `public` keywords

File: riposte-service-registration-eureka/src/main/java/com/nike/riposte/serviceregistration/eureka/helpers/EurekaVipAddressRoundRobinWithAzAffinityService.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    /**
+     * Round-robins the instances in the current availability zone returned by
+     * {@link #getActiveInstanceInfosForVipAddressBlocking(String)} to determine the *next* active {@link InstanceInfo}
+     * that should be called for the given VIP name. If there are no instances in the current availability zone, this
+     * method will round-robin the instances in all availability zones.
+     *
+     * @return The *next* active {@link InstanceInfo} that should be called for the given VIP name (using a round-robin
+     * strategy with current availability zone affinity), or an empty Optional if no active instances were found.
+     */
+    public Optional<InstanceInfo> getActiveInstanceInfoForVipAddressBlocking(String vipName) {
```
[the remark this answers is not in the document]
Comment (cjha) [reply riposte-service-registration-eureka/src/main/java/com/nike/riposte/serviceregistration/eureka/helpers/EurekaVipAddressRoundRobinWithAzAffinityService.java]: Yup, I'll fix it.

File: riposte-service-registration-eureka/src/main/java/com/nike/riposte/serviceregistration/eureka/helpers/EurekaVipAddressRoundRobinWithAzAffinityService.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+import com.netflix.discovery.DiscoveryClient;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+/**
+ * Service that uses the Eureka discovery client to provide all the instances associated with a VIP name ({@link
```
[the remark this answers is not in the document]
Comment (cjha) [reply riposte-service-registration-eureka/src/main/java/com/nike/riposte/serviceregistration/eureka/helpers/EurekaVipAddressRoundRobinWithAzAffinityService.java]: I'll clarify it.

**Comment by cjha** [comment]
I uploaded a new commit addressing the comments.

**Comment by nicmunroe** [comment]
Great, thank you @cjha this is a really good feature!

**Pull request merged**